### PR TITLE
feat: Course Level Error Handling for Empty States

### PR DESCRIPTION
**Pull request by shafqat-muneer**
[LEARNER-10020](https://2u-internal.atlassian.net/browse/LEARNER-10020): iOS - Course Level Error Handling for Empty States

### Portrait Mode:
| Light | Dark | Light | Dark |
|----------|----------|----------|----------|
| Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 41 25 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 51 01 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 42 00 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 49 45 |
| Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 42 09 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 50 03 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 42 22 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 50 20 |
| Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 42 31 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 50 26 |

### Landscape Mode:
| Light | Dark | Light | Dark |
|----------|----------|----------|----------|
| Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 41 28 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 51 03 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 42 02 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 49 58 |
| Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 42 10 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 50 05 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 42 24 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 50 22 |
| Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 42 32 | Simulator Screenshot - iPhone 15 - 2024-07-02 at 14 50 27 |

### CODE REVIEW

**Review by rnr — APPROVED** [review]

File: Course/Course/Presentation/Container/CourseContainerViewModel.swift [hunk 1]
```diff
@@ -252,15 +252,12 @@ public class CourseContainerViewModel: BaseCourseViewModel {
...
             isShowProgress = false
             isShowRefresh = false
             shouldShowUpgradeButton = false
-            if error.isInternetError || error is NoCachedDataError {
-                errorMessage = CoreLocalization.Error.slowOrNoInternetConnection
-            } else {
-                errorMessage = CoreLocalization.Error.unknownError
-            }
+            courseStructure = nil
+            courseVideosStructure = nil
```
Comment (saeedbashir) [inline Course/Course/Presentation/Container/CourseContainerViewModel.swift:260]: I guess we don't need to set these nil explicitly.

File: Course/Course/Presentation/Dates/CourseDatesViewModel.swift [hunk 2]
```diff
@@ -112,18 +112,14 @@ public class CourseDatesViewModel: ObservableObject {
...
             addCourseEventsIfNecessary()
-        } catch let error {
+        } catch {
             isShowProgress = false
-            if error.isInternetError || error is NoCachedDataError {
-                errorMessage = CoreLocalization.Error.slowOrNoInternetConnection
-            } else {
-                errorMessage = CoreLocalization.Error.unknownError
-            }
+            courseDates = nil
```
Comment (saeedbashir) [inline Course/Course/Presentation/Dates/CourseDatesViewModel.swift:122]: I guess we don't need to set it nil explicitly.

File: Course/Course/Presentation/Handouts/HandoutsView.swift [hunk 3]
```diff
@@ -107,6 +107,22 @@ struct HandoutsView: View {
...
+                    }
+                    .padding(.bottom, viewModel.connectivity.isInternetAvaliable
+                             ? 0 : OfflineSnackBarView.height)
+                    .transition(.move(edge: .bottom))
+                    .onAppear {
+                        doAfter(Theme.Timeout.snackbarMessageLongTimeout) {
+                            viewModel.errorMessage = nil
+                        }
+                    }
+                }
```
Comment (saeedbashir) [inline Course/Course/Presentation/Handouts/HandoutsView.swift]: I guess this isn't required and you can remove `showError` from `HandoutsViewModel`.

File: Course/Course/Presentation/Container/CourseContainerViewModel.swift (hunk 1 above)
Comment (shafqat-muneer) [reply Course/Course/Presentation/Container/CourseContainerViewModel.swift:260]: These properties are published, and based on these flags, a full-screen error view will appear. Therefore, they are necessary here.

File: Course/Course/Presentation/Dates/CourseDatesViewModel.swift (hunk 2 above)
Comment (shafqat-muneer) [reply Course/Course/Presentation/Dates/CourseDatesViewModel.swift:122]: Same [here](https://github.com/edx/edx-mobile-marketplace-ios/pull/28#discussion_r1686203011)

**Review by saeedbashir — APPROVED** [review]
LGTM 👍

**Comment by saeedbashir** [comment]
@shafqat-muneer Tests are failing otherwise PR looks good, please fix test cases before merging.

**Pull request merged**